New exception handler adoption with fallback

kmiller68 opened the pull request:
#### 66f743ca6ca6bdbf5241994991b09c8aba0c1c06
<pre>
New exception handler adoption with fallback
<a href="https://bugs.webkit.org/show_bug.cgi?id=271489">https://bugs.webkit.org/show_bug.cgi?id=271489</a>
<a href="https://rdar.apple.com/119256062">rdar://119256062</a>

Reviewed by Per Arne Vollan.

Patch by Brandon Kayes and Keith Miller.

This adopts the new hardened mach exception mechanism which was
introduced in darwin recently. See the radar for more details.
This attempt to land this patch should work with stale sandbox
profiles, which was causing problems with the previous iteration.

* Source/JavaScriptCore/jsc.cpp:
(main):
(CommandLine::parseArguments):
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/MachineContext.h:
(JSC::MachineContext::setInstructionPointer):
* Source/JavaScriptCore/runtime/VMTraps.cpp:
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
(JSC::Wasm::MachExceptionSigningKey::MachExceptionSigningKey):
(JSC::Wasm::trapHandler):
(JSC::Wasm::activateSignalingMemory):
(): Deleted.
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformMac.cmake:
* Source/WTF/wtf/PlatformRegisters.cpp:
(WTF::threadStatePCInternal):
* Source/WTF/wtf/PlatformRegisters.h:
* Source/WTF/wtf/spi/darwin/SandboxSPI.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::initMachExceptionHandlerThread):
(WTF::setExceptionPorts):
(WTF::activateSignalHandlersFor):
(WTF::finalizeSignalHandlers):
(WTF::toMachMask): Deleted.
* Source/WTF/wtf/threads/Signals.h:
(WTF::toMachMask):
(WTF::initializeSignalHandling):
(WTF::disableSignalHandling):
* Source/WTF/wtf/win/SignalsWin.cpp:
(WTF::finalizeSignalHandlers):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Tools/TestWebKitAPI/Tests/WTF/Signals.cpp:
(TEST):

Canonical link: <a href="https://commits.webkit.org/276579@main">https://commits.webkit.org/276579@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bce06d24f14abfbafdce706d09cc702133604c5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45016 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47519 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47674 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41024 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47322 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28206 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21525 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45594 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21184 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/38770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18035 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18601 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39901 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3064 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/38217 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41285 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40205 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49351 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/44473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/19989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16529 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21305 "Built successfully") | | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42719 "Found 2 new API test failures: /WebKitGTK/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns, /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51644 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6268 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20984 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/10501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1822,24 +1822,48 @@
+    (allow mach-message-send (kernel-mig-routine thread_set_exception_ports))
+#endif
+)
+
 (if (and (equal? (param "ENABLE_SANDBOX_MESSAGE_FILTER") "YES") (defined? 'mach-kernel-endpoint))
     (allow mach-kernel-endpoint
         (apply-message-filter
             (deny mach-message-send)
 #if ENABLE(LOCKDOWN_MODE_TELEMETRY)
             (with-filter (require-not (lockdown-mode))
                 (allow mach-message-send (kernel-mig-routines-blocked-in-lockdown-mode))
-                (allow mach-message-send (kernel-mig-routines-blocked-in-lockdown-mode-avoid-telemetry)))
+                (allow mach-message-send (kernel-mig-routines-blocked-in-lockdown-mode-avoid-telemetry))
+                (allow-thread-adopt-exception-handler))
             (with-filter (lockdown-mode)
                 (deny mach-message-send (with telemetry) (kernel-mig-routines-blocked-in-lockdown-mode))
                 (deny mach-message-send (with telemetry) (kernel-mig-routines-blocked-in-lockdown-mode-avoid-telemetry)))
 #else
             (allow mach-message-send (kernel-mig-routines-blocked-in-lockdown-mode))
             (allow mach-message-send (kernel-mig-routines-blocked-in-lockdown-mode-avoid-telemetry))
+            (allow-thread-adopt-exception-handler)
 #endif
 #if ENABLE(BLOCK_SET_EXCEPTION_PORTS)
             (with-filter (require-not (webcontent-process-launched))
-                (allow mach-message-send (kernel-mig-routine thread_set_exception_ports)))
+                (allow-mach-exceptions))
+#else
+            (allow-mach-exceptions)
```
I think this should only be allowed if not in Lockdown mode.

pvollan replied:
We can also do this in a follow-up if you prefer that.

kmiller68 replied:
Yeah, I think we could clean a bunch of this up. I was just going with what we had from before.

pvollan reviewed (commented):
The sandbox changes look good to me. We can harden the LDM sandbox on macOS even further. I added a comment about that.

bkayes commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -307,11 +361,21 @@ inline void setExceptionPorts(const AbstractLocker& threadGroupLocker, Thread& t
+
+#if CPU(ARM64) && HAVE(HARDENED_MACH_EXCEPTIONS)
+    // If we are a translated process in rosetta or failed to set up a hardened handler, use the old exception style
+    if (!WTF::isX86BinaryRunningOnARM() && !fallbackToOldExceptions.loadRelaxed()) {
```
(suggestion; non-blocking?) Maybe we should check `isX86BinaryRunningOnARM` above when we initially set `fallbackToOldExceptions`

pvollan commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3970,7 +3972,19 @@ void CommandLine::parseArguments(int argc, char** argv)
+            mask |= toMachMask(Signal::AccessFault);
+            mask |= toMachMask(Signal::FloatingPoint);
+            mask |= toMachMask(Signal::Breakpoint);
+#if !OS(DARWIN)
+            mask |= toMachMask(Signal::Abort);
+#endif // !OS(DARWIN)
```
Does HAVE(MACH_EXCEPTIONS) imply OS(DARWIN)? Perhaps this can be removed?

kmiller68 replied:
I think in theory there could be BSD distro's that have Mach. We can just leave it for now.

bkayes commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1822,24 +1822,48 @@
+    (when (defined? 'thread_adopt_exception_handler)
+        (allow mach-message-send (kernel-mig-routine thread_adopt_exception_handler))
+    )
+)
```
@pvollan just double checking this works properly on old and new SDK versions given it doesn't have the `#if HAVE(HARDENED_MACH_EXCEPTIONS)`

kmiller68 replied:
I think it should be ok as that won't be defined.

pvollan reviewed (approved):
R=me. I have reviewed the changes from https://github.com/WebKit/WebKit/pull/24761, which was already approved, and this looks good to me.

kmiller68 committed:
New exception handler adoption with fallback
https://bugs.webkit.org/show_bug.cgi?id=271489
rdar://119256062

Reviewed by Per Arne Vollan.

Patch by Brandon Kayes and Keith Miller.

This adopts the new hardened mach exception mechanism which was
introduced in darwin recently. See the radar for more details.
This attempt to land this patch should work with stale sandbox
profiles, which was causing problems with the previous iteration.

* Source/JavaScriptCore/jsc.cpp:
(main):
(CommandLine::parseArguments):
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
(JSC::initialize):
* Source/JavaScriptCore/runtime/MachineContext.h:
(JSC::MachineContext::setInstructionPointer):
* Source/JavaScriptCore/runtime/VMTraps.cpp:
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
(JSC::Wasm::MachExceptionSigningKey::MachExceptionSigningKey):
(JSC::Wasm::trapHandler):
(JSC::Wasm::activateSignalingMemory):
(): Deleted.
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformMac.cmake:
* Source/WTF/wtf/PlatformRegisters.cpp:
(WTF::threadStatePCInternal):
* Source/WTF/wtf/PlatformRegisters.h:
* Source/WTF/wtf/spi/darwin/SandboxSPI.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::initMachExceptionHandlerThread):
(WTF::setExceptionPorts):
(WTF::activateSignalHandlersFor):
(WTF::finalizeSignalHandlers):
(WTF::toMachMask): Deleted.
* Source/WTF/wtf/threads/Signals.h:
(WTF::toMachMask):
(WTF::initializeSignalHandling):
(WTF::disableSignalHandling):
* Source/WTF/wtf/win/SignalsWin.cpp:
(WTF::finalizeSignalHandlers):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Tools/TestWebKitAPI/Tests/WTF/Signals.cpp:
(TEST):

Canonical link: https://commits.webkit.org/276579@main
